core/src/worker/workflow/managed_run.rs

```diff
@@ -570,9 +578,36 @@ impl ManagedRun {
                 )
             }
         } else if should_report {
-            ActivationCompleteOutcome::ReportWFTFail(FailedActivationWFTReport::Report(
-                tt, cause, failure,
-            ))
+            // Check if we should fail the workflow instead of the WFT because of user's preferences
+            if matches!(cause, WorkflowTaskFailedCause::NonDeterministicError)
```

Comment:
There is a `self.metrics.wf_task_failed` above, can we remove this before that or move that into the conditionals? Similarly, will this make sure to increment the workflow failure counter? Any other things about workflow failure that is accounted for by the other workflow failure code path but not this one (e.g. eviction)?

Reply:
Eviction will happen on this path too.

As for the metric, I didn't avoid it since you could in some sense still consider this "failing" the task, it's just reported differently. I suppose we could go either way there. For the workflow failure counter, yeah I'll need to make sure that does get hit, thanks for mentioning.

Reply:
I have made it go through the same path so we get this stuff without much change.

I've kept it adding to the task fail metric. To my mind this is _both_ failing a task, and the workflow, in some sense. We already increment the task failure counter for WFT fails that aren't reported to server because they've happened >2 times, and this feels similar to that.

Reply:
Regarding task vs workflow fail, I think it's either/or and we should increment the counter for which path is taken instead of both. Inability to report a task failure I think may be different (TIL that we don't just send a task failure to the server every time the server sends us a task).

I think it's clearest to think from a user POV. When they opt in to this, e.g. `Workflow(FailureExceptionTypes=new[] { typeof(NondeterminismException), typeof(MyException) })`, they are basically flipping both of those from task fail to wf fail, even if internally we are only sending nondet flip to core and leaving user-exception flip in lang. So from a user POV, they are just two things that can happen that used to increment task fail and now increment wf fail metrics.

Reply:
Sure, I just think both interpretations are valid, but I'll make it either or since I don't care much